File: Study Plan/Top Interview 150/125 Valid Palindrome/ValidPalindrome.cs
/*  
    Problem
    --------------------------------------------------------------------------------
    125. Valid Palindrome
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    https://leetcode.com/problems/valid-palindrome/solutions/4384299/clean-solution-in-c-beats-80-01-in-runtime/

    - Runtime 66 ms (Beats 80.57% of users with C#)
    - Memory 42.44 MB (Beats 34.19% of users with C#)
*/
public class Solution
{
    public bool IsPalindrome(string phrase)
    {
        var characters = phrase
            .ToLower()
            .Where(character => char.IsLetterOrDigit(character)
                && !char.IsPunctuation(character)
                && !char.IsWhiteSpace(character))
            .ToArray();

        var forward = new string(characters);
        var backward = new string(forward.ToCharArray().Reverse().ToArray());

        return forward == backward;
    }
}

File: Study Plan/Top Interview 150/125 Valid Palindrome/ValidPalindrome.java
/*  
    Problem
    --------------------------------------------------------------------------------
    125. Valid Palindrome
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    - Runtime 4 ms (Beats 54.95% of users with Java)
    - Memory 43.11 MB (Beats 56.69% of users with Java)
*/

class Solution {
    public boolean isPalindrome(String s) {
        StringBuilder characters = new StringBuilder();
        for (int index = 0; index < s.length(); index++) {
            char character = s.charAt(index);
            if (Character.isLetterOrDigit(character)) {
                characters.append(Character.toLowerCase(character));
            }
        }
        String forward = characters.toString();
        String backward = new StringBuilder(forward).reverse().toString();
        return forward.equals(backward);
    }
}